chat-plugins/mafia.js

```diff
@@ -1,919 +1,931 @@
+					role.memo = role.memo.concat(MafiaData.alignments[id].memo);
+					if (!role.image && MafiaData.alignments[id].image) role.image = MafiaData.alignments[id].image;
+					target.splice(j, 1);
+					j--;
+				}
+			}
+			if (!role.alignment) role.alignment = 'town'; // Default to town
+			if (alignments.indexOf(role.alignment) === -1) alignments.push(role.alignment);
+			// Handle anything that is unknown
+			if (target.length) {
+				role.memo.push(`To learn more about your role, PM the host (${this.host}).`);
+			}
+			this.originalRoles.push(role);
+		}
+		if (alignments.length < 2) problems.push(`There must be at least 2 different alignments in a game!`);
+		if (problems.length) {
+			this.originalRoles = [];
+			return problems;
+		}
+		this.roles = this.originalRoles.slice();
+		this.originalRoleString = this.originalRoles.slice().map(r => { return `<span style="font-weight:bold;color:${MafiaData.alignments[r.alignment].color || '#FFF'}">${r.name}</span>`; }).join(', ');
+		this.roleString = this.originalRoleString;
+		this.updatePlayers();
+		return [];
+	}
+
+	start(user) {
+		if (!user) return false;
+		if (this.phase !== 'locked') {
+			if (this.phase === 'signups') return user.sendTo(this.room, `You need to close the signups first.`);
+			return user.sendTo(this.room, `The game is already started!`);
+		}
+		if (this.playerCount < 2) return user.sendTo(this.room, `You need at least 2 players to start.`);
+		if (!Object.keys(this.roles).length) return user.sendTo(this.room, `You need to set the roles before starting.`);
+		if (Object.keys(this.roles).length !== this.playerCount) return user.sendTo(this.room, `You have not provided enough roles for the players.`);
+		this.started = true;
+		this.played = Object.keys(this.players);
+		this.sendRoom(`The game of ${this.title} is starting!`, {declare: true});
+		this.played.push(this.hostid);
+		this.distributeRoles();
+		this.day(false, true);
+	}
+
+	distributeRoles() {
+		if (this.phase !== 'locked' || !Object.keys(this.roles).length) return false;
+		this.sendRoom(`The roles are being distributed...`);
+		let roles = Dex.shuffle(this.roles.slice());
+		let alignments = {};
+		for (let p in this.players) {
+			let role = roles.shift();
+			this.players[p].role = role;
+			if (!['town', 'solo'].includes(role.alignment)) {
+				if (!alignments[role.alignment]) alignments[role.alignment] = [];
+				alignments[role.alignment].push(this.players[p].userid);
+			}
+		}
+		for (let a in alignments) {
+			for (let p of alignments[a]) {
+				this.players[p].alliedPlayers = alignments[a];
+			}
+		}
+		this.sendRoom(`The roles have been distributed.`, {declare: true});
+		this.updatePlayers();
+		return true;
 	}
 
-	updateTarget(image) {
-		if (this.target) {
-			this.targetWindow(image, 'Targeting ' + Chat.escapeHTML(this.target.name) + '!<br />', true);
-		} else {
-			this.targetWindow(image, 'You chose to not target anybody.<br />', true);
+	getPartners(alignment, player) {
```

Comment:
Can't this just use ``player.alliedPlayers``?

Reply:
I did it this way so it updates when partners die.